lib/redis.rb

```diff
@@ -2130,13 +2130,14 @@ def unwatch
 
   def pipelined
     synchronize do |client|
+      pipeline = Pipeline.new
       begin
-        original, @client = @client, Pipeline.new
+        original, @client = @client, pipeline
         yield(self)
-        original.call_pipeline(@client)
       ensure
         @client = original
       end
+      original.call_pipeline(pipeline)
```

Comment:
> I would have liked if instead of modifying the client directly, we would have an API like:

We were talking about that with @davidcornu, and definitely that would be a huge improvement. Maybe that can be worked out in a backward compatible way with `Proc#arity`:

```ruby
def pipelined(&block)
  if block.arity == 1
    yield pipeline
  else
    @client = pipeline
    yield
    ....
  end
end
```